File: guests/java/dev/kaya/milestone2kt/A11y.java
package dev.kaya.milestone2kt;

import dev.kaya.KayaApp;

/**
 * The accessibility conformance scene from the JVM: the two universal
 * props (a11yId, a11yLabel), read back out of the platform's own
 * accessibility tree. Exactly ONE container of each container kind:
 * container targets are ordinal. See guests/rust/a11y.rs;
 * tools/scenes/a11y.steps.
 */
final class A11y {
    static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            tx.mount(tx.column(() -> {
                // Caption-bearing controls: deliberately NOT labelled,
                // so the platform has to speak the caption.
                tx.button("Save").a11yId("save").a11yHint("save the draft");
                tx.checkbox("Details", null).a11yId("details")
                        .a11yHint("show more detail");
                tx.button("Reset").a11yId("reset");
                tx.label("Ready").a11yId("status");
                // Caption-less controls: an app MUST name these.
                tx.entry().a11yId("name").a11yLabel("Full name");
                tx.textarea().a11yId("notes").a11yLabel("Notes");
                tx.slider(0.0, 1.0, 0.5, null).a11yId("volume").a11yLabel("Volume");
                tx.progress(0.25).a11yId("loading").a11yLabel("Loading");
                // THE MARK THE APP'S OWN BUILD SHIPPED: the bytes never
                // enter this JVM's heap, and try-with-resources is safe
                // because the blob table already holds its own
                // reference.
                try (KayaApp.Asset mark = KayaApp.asset("images/a11y-logo.png")) {
                    tx.image(mark).a11yId("logo").a11yLabel("Logo");
                }
                tx.select(new String[] {"Red", "Green"}, 0, null)
                        .a11yId("color").a11yLabel("Color");
                tx.radio(new String[] {"Small", "Large"}, 0, null)
                        .a11yId("size").a11yLabel("Size");
                tx.grid(2, () -> {
                    tx.label("Name");
                    tx.label("Ada");
                }).a11yId("cells").a11yLabel("Cells");
                tx.scroll(() -> tx.label("Item")).a11yId("feed").a11yLabel("Feed");
                tx.row(() -> {
                    tx.button("Cancel").a11yId("cancel");
                    tx.button("OK").a11yId("ok");
                }).a11yId("actions").a11yLabel("Actions");
                // A spoken name that FOLLOWS A SIGNAL: the live trio's
                // Signal overloads, the template zone's shape live.
                KayaApp.Signal<String> spoken = tx.signal("Before");
                tx.label("Spoken").a11yId("spoken").a11yLabel(spoken);
                tx.button("Rename", inner -> inner.write(spoken, "After"))
                        .a11yId("rename");
            }).a11yId("form").a11yLabel("Form"));
            return null;
        });

        app.dispatchLoop();
    }
}

File: guests/go/a11y/a11y.go
// The a11y conformance scene (tools/scenes/a11y.steps). Keep exactly ONE
// container of each kind: container targets are ordinal.
package a11y

import (
	kaya "dev.kaya/bindings/go"
)

func App() *kaya.App {
	app := kaya.NewApp()

	app.Build(func(tx *kaya.Tx) {
		tx.Mount(tx.Column(func() {
			// Deliberately NOT labelled: the platform speaks the caption.
			tx.Button("Save", nil).A11yID("save").A11yHint("save the draft")
			tx.Checkbox("Details", nil).A11yID("details").A11yHint("show more detail")
			tx.Button("Reset", nil).A11yID("reset")
			tx.LabelText("Ready").A11yID("status")
			tx.Entry(nil).A11yID("name").A11yLabel("Full name")
			tx.Textarea(nil).A11yID("notes").A11yLabel("Notes")
			tx.Slider(0.0, 1.0, 0.5, nil).A11yID("volume").A11yLabel("Volume")
			tx.Progress(0.25).A11yID("loading").A11yLabel("Loading")
			logo := tx.Asset("images/a11y-logo.png")
			tx.ImageAsset(logo).A11yID("logo").A11yLabel("Logo")
			tx.Select([]string{"Red", "Green"}, 0, nil).
				A11yID("color").A11yLabel("Color")
			tx.Radio([]string{"Small", "Large"}, 0, nil).
				A11yID("size").A11yLabel("Size")
			tx.Grid(2, func() {
				tx.LabelText("Name")
				tx.LabelText("Ada")
			}).A11yID("cells").A11yLabel("Cells")
			tx.Scroll(func() {
				tx.LabelText("Item")
			}).A11yID("feed").A11yLabel("Feed")
			tx.Row(func() {
				tx.Button("Cancel", nil).A11yID("cancel")
				tx.Button("OK", nil).A11yID("ok")
			}).A11yID("actions").A11yLabel("Actions")
			spoken := tx.Signal("Before")
			tx.LabelText("Spoken").A11yID("spoken").BindA11yLabel(spoken)
			tx.Button("Rename", func(tx *kaya.Tx) {
				tx.Write(spoken, "After")
			}).A11yID("rename")
		}).A11yID("form").A11yLabel("Form"))
	})

	return app
}
